test/common/TestCommon.hs

```diff
@@ -0,0 +1,179 @@
+  quad x y
+
+listsToParameters :: ([Double], [Double]) -> Domains Double
+listsToParameters (a0, a1) =
+  (V.fromList a0, V.singleton $ V.fromList a1, V.empty, V.empty)
+
+listsToParameters4 :: ([Double], [Double], [Double], [Double]) -> Domains Double
+listsToParameters4 (a0, a1, a2, aX) =
+  ( V.fromList a0
+  , V.singleton $ V.fromList a1
+  , if null a2 then V.empty else V.singleton $ HM.matrix 1 a2
+  , if null aX then V.empty else V.singleton $ OT.fromList [length aX] aX )
+
+quickCheckTest0 :: TestName
+       -> (forall d r m. ( DualMonad d r m
+                         , Floating (Out (DualNumber d (Vector r)))
+                         , Floating (Out (DualNumber d (OS.Array '[2] r))) )
+           => DualNumberVariables d r -> m (DualNumber d r))
+       -> ((Double, Double, Double) -> ([Double], [Double], [Double], [Double]))
+       -> TestTree
+quickCheckTest0 txt f fArg =
+  qcTestRanges txt f (listsToParameters4 . fArg) ((-2, -2, -2), (2, 2, 2)) ((-1e-7, -1e-7, -1e-7), (1e-7, 1e-7, 1e-7)) (-10, 10)
+
+-- A quick check to compare the derivatives and values of 2 given functions.
+cmpTwo :: (forall d r m. ( DualMonad d r m
+                         , r ~ Double
+                         , Floating (Out (DualNumber d (Vector r)))
+                         , Floating (Out (DualNumber d (OS.Array '[2] r))) )
+           => DualNumberVariables d r -> m (DualNumber d r))
+       ->     (forall d r m. ( DualMonad d r m
+                         , r ~ Double
+                         , Floating (Out (DualNumber d (Vector r)))
+                         , Floating (Out (DualNumber d (OS.Array '[2] r))) )
+           => DualNumberVariables d r -> m (DualNumber d r))
+       -> Domains Double
+       -> Domains Double
+       -> Domains Double
+       -> Domains Double
+       -> Property
+cmpTwo f1 f2 params1 params2 ds1 ds2 =
+  close2 (dFastForward f1 params1 ds1) (dFastForward f2 params2 ds2)
+
+-- A quick check to compare the derivatives and values of 2 given functions.
+cmpTwoSimple :: (forall d r m. ( DualMonad d r m
+                         , r ~ Double
+                         , Floating (Out (DualNumber d (Vector r)))
+                         , Floating (Out (DualNumber d (OS.Array '[2] r))) )
+                 => DualNumberVariables d r -> m (DualNumber d r))
+       ->     (forall d r m. ( DualMonad d r m
+                         , r ~ Double
+                         , Floating (Out (DualNumber d (Vector r)))
+                         , Floating (Out (DualNumber d (OS.Array '[2] r))) )
+           => DualNumberVariables d r -> m (DualNumber d r))
+       -> Domains Double
+       -> Domains Double
+       -> Property
+cmpTwoSimple f1 f2 parameters ds =
+  cmpTwo f1 f2 parameters parameters ds ds
+
+-- A quick consistency check of all the kinds of derivatives and gradients
+-- and all kinds of computing the value of the objective function.
+qcPropDom :: (forall d r m. ( DualMonad d r m
+                         , r ~ Double
```

Comment:
broken formatting again?

Reply:
Throughout the codebase I have seen different ways of formatting this kind of stuff, for example [here](https://github.com/Mikolaj/horde-ad/blob/91dc3840ee27cc75ce1b7c5a1411600eb66bfe94/src/HordeAd/Core/Engine.hs#L246) or [here](https://github.com/Mikolaj/horde-ad/blob/91dc3840ee27cc75ce1b7c5a1411600eb66bfe94/test/common/TestSimpleDescent.hs#L64) or [here](https://github.com/Mikolaj/horde-ad/blob/91dc3840ee27cc75ce1b7c5a1411600eb66bfe94/test/common/TestSimpleDescent.hs#L185). What is the right way? It is enough if you point me to the right example.

Reply:
There's lots of broken formatting in this codebase from blind query-replace. I'm sorry about that. The rule is that `=>` and `->` align with `::`, unless they are inside parens, in which case they start in the column one higher than where the starting paren is. So all the three examplex are fine. The first breaks line after a long function name and starts the new line with `::`, the second doesn't break line after the forall phrase, the third is the "inside parens" case. Another rule is that if a line starts with a comma, it aligns with the paren above and is followed by a space, just as the paren.

Reply:
Yet a separate case are perhaps short argument names, like [here](https://github.com/Mikolaj/horde-ad/blob/91dc3840ee27cc75ce1b7c5a1411600eb66bfe94/src/HordeAd/Core/Engine.hs#L301)? Did you think of enforcing those rules on the server side as part of CI/CD, or of implementing them as part of `haskell-mode`?

Reply:
You are right, the rule should read "`=>` and `->` that start a line align...". Yes, I think the new haskell mode (not the decade old version I'm using) may be customizable enough to support such a rule. Or you can just hack some emacs-lisp. :) A really tempting alternative is ormolu or another auto-formatter. It prevents formatting getting broken due to `sed` query-replace. However, I have yet to see auto-formatting that doesn't make me cringe and also auto-formatters can't cope with 80-characters rule, in part because the rule has to admit exceptions, e.g., for very long strings (or you need to refactor code to split the strings; auto-formatter can't do that).